engine/runtime/src/main/java/org/enso/interpreter/runtime/state/ExecutionEnvironment.java

```diff
@@ -22,80 +18,48 @@ public class ExecutionEnvironment {
-    this.keys = keys;
     this.permissions = permissions;
   }
 
   public String getName() {
     return this.name;
   }
 
-  public ExecutionEnvironment withContextEnabled(Atom context) {
-    return update(context, true);
-  }
-
-  public ExecutionEnvironment withContextDisabled(Atom context) {
-    return update(context, false);
-  }
-
-  private ExecutionEnvironment update(Atom context, boolean value) {
-    assert context.getConstructor().getType()
-        == EnsoContext.get(null).getBuiltins().context().getType();
-    int keyFound = -1;
-    for (int i = 0; i < keys.length; i++) {
-      if (keys[i].equals(context.getConstructor().getName())) {
-        keyFound = i;
-      }
-    }
-    String[] keys1;
-    Boolean[] permissions1;
-    if (keyFound != -1) {
-      keys1 = cloneArray(keys, new String[keys.length]);
-      permissions1 = cloneArray(permissions, new Boolean[keys.length]);
-      permissions1[keyFound] = value;
-    } else {
-      keys1 = cloneArray(keys, new String[keys.length + 1]);
-      permissions1 = cloneArray(permissions, new Boolean[keys.length + 1]);
-      keyFound = keys.length;
-      keys1[keyFound] = context.getConstructor().getName();
-      permissions1[keyFound] = value;
-    }
-    return new ExecutionEnvironment(name, keys1, permissions1);
-  }
-
-  private <T> T[] cloneArray(T[] fromArray, T[] toArray) {
-    for (int i = 0; i < fromArray.length; i++) {
-      toArray[i] = fromArray[i];
-    }
-    return toArray;
-  }
-
-  public Boolean hasContextEnabled(String context) {
-    int keyFound = -1;
-    for (int i = 0; i < keys.length; i++) {
-      if (keys[i].equals(context)) {
-        keyFound = i;
-      }
-    }
-    if (keyFound != -1) {
-      return permissions[keyFound];
+  /**
+   * Returns copy of this {@link ExecutionEnvironment} with new permissions for the given context.
+   *
+   * @param ctor Constructor of {@code Standard.Base.Runtime.Context} type.
+   * @param contextBuiltin The builtin type of {@code Standard.Base.Runtime.Context}.
+   * @param value The new value for the permissions.
+   * @return A copy of the execution environment with the new permissions.
+   */
+  ExecutionEnvironment update(AtomConstructor ctor, Context contextBuiltin, boolean value) {
+    assert ctor.getType() == contextBuiltin.getType();
+    ContextPermissions newPermissions;
+    if (ctor == contextBuiltin.getInput()) {
+      newPermissions =
+          new ContextPermissions(value, permissions.output(), permissions.dataflowStacktrace());
+    } else if (ctor == contextBuiltin.getOutput()) {
+      newPermissions =
+          new ContextPermissions(permissions.input(), value, permissions.dataflowStacktrace());
+    } else if (ctor == contextBuiltin.getDataflowStackTrace()) {
+      newPermissions = new ContextPermissions(permissions.input(), permissions.output(), value);
     } else {
-      return false;
+      throw CompilerDirectives.shouldNotReachHere("Unknown context `" + ctor.getName() + "`");
```

Comment:
`raiseAssertionError` please.

Reply:
Done in ee54f74d6b7d4d230c63e90f2fc83b41e955a401